File: 03_Backend/app/schemas/receipt.py
"""Receipt API contracts."""

from datetime import date, datetime

from pydantic import BaseModel, Field

from app.schemas.base import BaseResponse
from app.schemas.pagination import PaginationParams


class ReceiptResponse(BaseResponse):
    numero: str
    repas_uuid: str
    utilisateur_uuid: str
    nom: str
    prenom: str
    matricule: str | None = None
    type_utilisateur: str
    categorie_uuid: str
    categorie_nom: str
    type_identification: str
    date_repas: date
    heure_repas: datetime
    nombre_impressions: int = 0
    derniere_impression: datetime | None = None
    qr_token: str | None = None


class ReceiptQrRequest(BaseModel):
    token: str = Field(min_length=20, max_length=300)


class ReceiptFilterParams(PaginationParams):
    date_from: date | None = None
    date_to: date | None = None
    category: str | None = Field(None, max_length=100)
    user_type: str | None = Field(None, max_length=20)
    identification_type: str | None = Field(None, max_length=20)

File: mobile_app/lib/features/receipts/domain/entities/receipt.dart
class Receipt {
  final String uuid;
  final String number;
  final String mealUuid;
  final String userUuid;
  final String firstName;
  final String lastName;
  final String? employeeNumber;
  final String userType;
  final String categoryUuid;
  final String categoryName;
  final String identificationType;
  final DateTime mealDate;
  final DateTime servedAt;
  final int printCount;
  final DateTime? lastPrintedAt;
  final String? qrToken;

  const Receipt({
    required this.uuid,
    required this.number,
    required this.mealUuid,
    required this.userUuid,
    required this.firstName,
    required this.lastName,
    this.employeeNumber,
    required this.userType,
    required this.categoryUuid,
    required this.categoryName,
    required this.identificationType,
    required this.mealDate,
    required this.servedAt,
    this.printCount = 0,
    this.lastPrintedAt,
    this.qrToken,
  });

  String get fullName => '$firstName $lastName'.trim();

  factory Receipt.fromJson(Map<String, dynamic> json) => Receipt(
    uuid: json['uuid'] as String,
    number: json['numero'] as String,
    mealUuid: json['repas_uuid'] as String,
    userUuid: json['utilisateur_uuid'] as String,
    firstName: json['prenom'] as String? ?? '',
    lastName: json['nom'] as String? ?? '',
    employeeNumber: json['matricule'] as String?,
    userType: json['type_utilisateur'] as String? ?? '',
    categoryUuid: json['categorie_uuid'] as String? ?? '',
    categoryName: json['categorie_nom'] as String? ?? '',
    identificationType: json['type_identification'] as String? ?? '',
    mealDate: DateTime.parse(json['date_repas'] as String),
    servedAt: DateTime.parse(json['heure_repas'] as String),
    printCount: json['nombre_impressions'] as int? ?? 0,
    lastPrintedAt: json['derniere_impression'] == null
        ? null
        : DateTime.parse(json['derniere_impression'] as String),
    qrToken: json['qr_token'] as String?,
  );
}
